File: classroom_app/services/process_material_import_policy.py
from __future__ import annotations

import os
from typing import Any

from fastapi import HTTPException


PROCESS_DOCUMENT_IMPORT_ALLOWED_EXTENSIONS: tuple[str, ...] = (
    ".doc",
    ".docx",
    ".pdf",
    ".png",
    ".jpg",
    ".jpeg",
    ".webp",
    ".bmp",
    ".gif",
    ".md",
    ".txt",
)
PROCESS_DOCUMENT_IMPORT_ACCEPT = ",".join(PROCESS_DOCUMENT_IMPORT_ALLOWED_EXTENSIONS)
PROCESS_DOCUMENT_IMPORT_FORMAT_LABEL = "Word/PDF、Markdown/TXT 或常见图片"
PROCESS_DOCUMENT_IMPORT_MAX_FILES = 8
PROCESS_DOCUMENT_IMPORT_MAX_BYTES = 30 * 1024 * 1024


def validate_process_document_import_file_count(files: Any) -> int:
    count = len(files or [])
    if count <= 0:
        raise HTTPException(400, "请至少选择一个文件")
    if count > PROCESS_DOCUMENT_IMPORT_MAX_FILES:
        raise HTTPException(400, f"最多一次导入 {PROCESS_DOCUMENT_IMPORT_MAX_FILES} 个文件")
    return count


def normalize_process_import_filename(filename: Any, *, fallback: str = "file") -> str:
    normalized = os.path.basename(str(filename or "").strip())
    return normalized or fallback


def validate_process_document_import_filename(filename: Any, *, document_label: str = "过程材料") -> None:
    name = normalize_process_import_filename(filename)
    ext = os.path.splitext(name)[1].lower()
    if ext in PROCESS_DOCUMENT_IMPORT_ALLOWED_EXTENSIONS:
        return
    ext_label = ext or "无扩展名"
    raise HTTPException(
        415,
        f"{document_label}导入暂不支持 {ext_label} 文件，请上传{PROCESS_DOCUMENT_IMPORT_FORMAT_LABEL}。",
    )


def validate_process_document_import_file_bytes(data: bytes | bytearray, *, filename: Any) -> None:
    name = normalize_process_import_filename(filename)
    size = len(data or b"")
    if size <= 0:
        raise HTTPException(400, f"《{name}》是空文件，请重新选择。")
    if size > PROCESS_DOCUMENT_IMPORT_MAX_BYTES:
        raise HTTPException(400, f"《{name}》超过 30MB 单文件上限，请压缩或拆分后再导入。")

File: static/js/process_material_import_policy.js
export const PROCESS_DOCUMENT_IMPORT_EXTENSIONS = [
    '.doc',
    '.docx',
    '.pdf',
    '.png',
    '.jpg',
    '.jpeg',
    '.webp',
    '.bmp',
    '.gif',
    '.md',
    '.txt',
];

export const PROCESS_DOCUMENT_IMPORT_ACCEPT = PROCESS_DOCUMENT_IMPORT_EXTENSIONS.join(',');
export const PROCESS_DOCUMENT_IMPORT_FORMAT_HINT = '支持 Word（doc/docx）、PDF、Markdown/TXT 和常见图片；不支持 Excel、压缩包或无扩展名文件。';
export const PROCESS_DOCUMENT_IMPORT_MAX_FILES = 8;
export const PROCESS_DOCUMENT_IMPORT_MAX_BYTES = 30 * 1024 * 1024;

function fileExtension(file) {
    const name = String(file?.name || '').trim();
    const index = name.lastIndexOf('.');
    return index > 0 ? name.slice(index).toLowerCase() : '';
}

export function getProcessDocumentImportFileKey(file) {
    if (!file) return '';
    const name = String(file.name || '').trim().toLowerCase();
    if (!name) return '';
    const size = Number(file.size || 0);
    const modified = Number(file.lastModified || 0);
    return `${name}:${size}:${modified}`;
}

export function formatProcessImportFileSize(bytes) {
    const size = Number(bytes || 0);
    if (!Number.isFinite(size) || size <= 0) return '0 KB';
    if (size >= 1024 * 1024) {
        return `${(size / 1024 / 1024).toFixed(size >= 10 * 1024 * 1024 ? 0 : 1)} MB`;
    }
    return `${Math.max(1, Math.round(size / 1024))} KB`;
}

export function getProcessDocumentImportFileProblem(file, currentCount = 0) {
    if (currentCount >= PROCESS_DOCUMENT_IMPORT_MAX_FILES) {
        return `一次最多导入 ${PROCESS_DOCUMENT_IMPORT_MAX_FILES} 个文件，请先移除部分文件。`;
    }
    if (!file) return '请选择要导入的文件。';
    const name = String(file.name || '未命名文件');
    const ext = fileExtension(file);
    if (!ext || !PROCESS_DOCUMENT_IMPORT_EXTENSIONS.includes(ext)) {
        return `不支持《${name}》的格式${ext ? `（${ext}）` : ''}，请上传 Word/PDF、Markdown/TXT 或图片。`;
    }
    if (Number(file.size || 0) <= 0) {
        return `《${name}》是空文件，请重新选择。`;
    }
    if (Number(file.size || 0) > PROCESS_DOCUMENT_IMPORT_MAX_BYTES) {
        return `《${name}》超过 30MB 单文件上限，请压缩或拆分后再导入。`;
    }
    return '';
}

export function getProcessDocumentImportDuplicateProblem(file, pickedFiles = []) {
    const key = getProcessDocumentImportFileKey(file);
    if (!key) return '';
    const hasDuplicate = Array.from(pickedFiles || []).some((picked) => getProcessDocumentImportFileKey(picked) === key);
    if (!hasDuplicate) return '';
    const name = String(file.name || '未命名文件');
    return `文件已在待导入列表中：${name}`;
}
